### ISSUE TITLE
About CLIP RN-50

### ISSUE THREAD

**Issue by ghost** [linyq2117/CLIP-ES#5]
Thank you very much for your exciting work! 

Would like to ask if you can provide the code for CLIP RN50 to generate CAMs?

I wish you good luck with your research and a happy life.

**Comment by linyq2117** [comment]
Thanks for your interest in our work.

This repo is mainly designed for CLIP-ViT. If you want to use RN50, you need to modify some lines as follows (take voc as example):
### generate_cams_voc12.py
#### 1. Comment out or remove line 134 https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L134
#### 2. Replace `image_features` with `image` in line 144: `input_tensor = [image, text_features_temp.to(device_id), h, w]` https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L144
#### 3. Comment out or remove line 160 - line 197, [line 202,](https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L202) [line 207,](https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L207) [line 213,](https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L213) uncomment [line 212,](https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L212)
#### 4. Change `target_layers` in line 239: `target_layers = [model.visual.layer4[-1]]` https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L239 
#### 5. Set `reshape_transform` to `None` in line 240: `cam = GradCAM(model=model, target_layers=target_layers, reshape_transform=None)` https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/generate_cams_voc12.py#L240 

### clip/model.py
#### 1. Add a forward function for resnet50 in CLIP object, such as:
```
def forward_features_resnet50(self, image, text_features, h, w):
        image_features = self.encode_image(image, h, w)
        # normalized features
        image_features = image_features / image_features.norm(dim=1, keepdim=True)
        text_features = text_features / text_features.norm(dim=1, keepdim=True)
        # cosine similarity as logits
        logit_scale = self.logit_scale.exp()
        logits_per_image = logit_scale * image_features @ text_features.t()

        # shape = [global_batch_size, global_batch_size]
        logits_per_image = logits_per_image.softmax(dim=-1)
        attn_weight = None

        return logits_per_image, attn_weight
```

### pytorch_grad_cam/activations_and_gradients.py
#### 1. Set `self.height=H // 32, self.width=W // 32` in line 40-41. https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/pytorch_grad_cam/activations_and_gradients.py#L40-L41
#### 2. Change forward function to `self.model.forward_features_resnet50(x[0], x[1], H, W)` in https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/pytorch_grad_cam/activations_and_gradients.py#L45

**Comment by ghost** [comment]
感谢您耐心的回复！

我按照上述步骤改了代码，可以正常运行。但是我在 CLIP-RN50 产生的 CAM 的基础上加上了 CRF ，结果比较低。
`
python eval_cam_with_crf.py --eval_only
`
得到的结果
`
'Pixel Accuracy': 0.7503926354038674, 'Mean Accuracy': 0.14167843883095207, 'Frequency Weighted IoU': 0.5694430970966254, 'Mean IoU': 0.127361159247873
`
请问这是什么原因造成的？（用ViT-B/16产生的伪标签精度和论文一致）

**Comment by linyq2117** [comment]
You can try different bg thresholds (such as 8 or 9 indicated by the result of `eval_cam.py` in https://github.com/linyq2117/CLIP-ES/blob/f0cecc74eb653f5f11abf4989f2463301009f155/eval_cam_with_crf.py#L111 We set it to 1 because it performs well on CLIP-ViT. For RN50, the best threshold may differ.

**Comment by ghost** [comment]
Thanks! It works.
